### ISSUE TITLE
Race between protocol Disconnect message and socket closing

### ISSUE THREAD

**Issue by nkostis** [gnatmq/gnatmq#22]
Handling a protocol disconnect (`MqttMsgDisconnect.MQTT_MSG_DISCONNECT_TYPE` in `DispatchEventThread()`) will result in `OnMqttMsgDisconnected()` which will call the `MqttMsgDisconnected` callback(s) and eventually close the client.

Handling a socket close will call `OnConnectionClosing()` which will call the `ConnectionClosed` callback(s) and also close the client. 

These 2 events can race each other, and in some cases you may get both callbacks called. Both callbacks by default call `MqttClient.Close()`, which does not protect against being called more than once.
(It sets IsConnected to false at its end, but it doesn't check it at its beginning. Moreover users are allowed to add their own callbacks, which results in varying levels of weirdness)

This is especially evident if `OnMqttMsgDisconnected()` takes a bit longer to execute and in the meantime the socket has (gracefully) dropped, so when it exits, it also executes the final part of `DispatchEventThread()` and re-calls `MqttClient.Close()`.

TL;DR or how i personally solve this:
1. i added protection against MqttClient.Close() being called twice
2. the race might be resolved by adding a lock on `MqttClient.isConnectionClosing`, but i just partitioned the `isConnectionClosing` part of `DiscpatchEventThread` into its own block which re-checks isRunning.

If anyone's interested or Paolo requests it, i can create pull requests for these, but for now [here's a simple patch](https://www.dropbox.com/s/xchr7mso9cy0qb0/gnatmq_nkostis_disconnect-race.patch?dl=0).

**Comment by dtflx01** [comment]
Hi nkostis,
Your correction seems interesting. Please can you post the complete file there ?
Thank you very much, regards

**Comment by dtflx01** [comment]
I applied the correction suggested by @nkostis.
See [https://github.com/dtflx01/gnatmq/commit/0bfb6b85d6473f03fb401520a6fb2cf4643d61cc](https://github.com/dtflx01/gnatmq/commit/0bfb6b85d6473f03fb401520a6fb2cf4643d61cc)

Please merge it.
Regards

**Comment by VladimirAkopyan** [comment]
@dtflx01  I see the correction, and that you've added me to your repo. However, could you please submit a pull request? It would be good to have a PR for every change to master. You don't require special right to do a PR.

**Comment by dtflx01** [comment]
OK, I'll do it @VladimirAkopyan .

**Comment by nkostis** [comment]
@dtflx01 heh.. i just saw this, you referenced me but without the mighty @ so i received nothing..!  glad to have helped months ago :)

**Comment by dtflx01** [comment]
@nkostis ;-) , thank you for your contribution.
